flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -5581,10 +5628,210 @@ class ArrayExprLowering {
+                           addSub(sub);
+                       }
+                       ty = fir::unwrapSequenceType(ty);
+                     },
+                     [&](const Fortran::evaluate::Component *x) {
+                       ty = addField(*x);
+                     }},
+                 v);
+    }
+    if (dim == 0) {
+      assert(ty.isa<fir::SequenceType>() && "must be an array to have rank");
+      pushAllIters();
+      ty = fir::unwrapSequenceType(ty);
+    }
+    return {result, ty};
+  }
+
+  /// Apply the reversed path components to the value returned from `load`.
+  CC applyPathToArrayLoad(mlir::Location loc, fir::ArrayLoadOp load) {
+    auto revPath = reversePath; // Force a copy to be made.
+    if (semant == ConstituentSemantics::ProjectedCopyInCopyOut) {
+      destination = load;
+      auto innerArg = explicitSpace->findArgumentOfLoad(load);
+      return [=](IterSpace iters) mutable {
+        auto [path, eleTy] = lowerPath(loc, revPath, load.getType(), iters);
+        auto toTy = adjustedArrayElementType(eleTy);
+        auto castedElement =
+            builder.createConvert(loc, toTy, iters.getElement());
+        auto update = builder.create<fir::ArrayUpdateOp>(
+            loc, innerArg.getType(), innerArg, castedElement, path,
+            load.typeparams());
+        // Flag the offsets as "Fortran" as they are not zero-origin.
+        update->setAttr(fir::factory::attrFortranArrayOffsets(),
+                        builder.getUnitAttr());
+        return arrayLoadExtValue(builder, loc, load, {}, update);
+      };
+    }
+    return [=](IterSpace iters) mutable {
+      auto [path, eleTy] = lowerPath(loc, revPath, load.getType(), iters);
+      auto resTy = adjustedArrayElementType(eleTy);
+      // TODO: When loading a subobject array, convert the fetch (value) to a
+      // fetch reference. This reference can then be used in a secondary
+      // array_load. Consider handling this another way.
+      if (semant == ConstituentSemantics::RefOpaque &&
+          !fir::isa_ref_type(resTy))
+        resTy = builder.getRefType(resTy);
+      auto fetch = builder.create<fir::ArrayFetchOp>(loc, resTy, load, path,
+                                                     load.typeparams());
+      // Flag the offsets as "Fortran" as they are not zero-origin.
+      fetch->setAttr(fir::factory::attrFortranArrayOffsets(),
+                     builder.getUnitAttr());
+      return arrayLoadExtValue(builder, loc, load, path, fetch);
+    };
+  }
+  CC applyPathToArrayLoad(fir::ArrayLoadOp load) {
+    auto loc = getLoc();
+    auto lambda = applyPathToArrayLoad(loc, load);
+    reversePath.clear();
+    return lambda;
+  }
+
+  CC genexp(const Fortran::semantics::Symbol &x) {
+    if (auto load = explicitSpace->findBinding(&x))
+      return applyPathToArrayLoad(load);
+    if (pathIsEmpty())
+      return [=](IterSpace) { return asScalar(x); };
+    return [=](IterSpace) { return ExtValue{}; };
```

Comment:
Should this last line be a fir::emitFatalError or are they cases where this can be reached and the empty ExtValue leads to a well defined behavior ?

Reply:
I'll double check. This may be vestigial.